network/server.go

```diff
@@ -237,70 +179,291 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
 		}
 
-		// start discovery
-		s.discovery = &discovery{srv: s}
+		// Write the networking private key to disk
+		if setErr := secretsManager.SetSecret(secrets.NetworkKey, libp2pKeyEncoded); setErr != nil {
+			return nil, fmt.Errorf("unable to store networking private key to Secrets Manager, %w", setErr)
+		}
 
-		// try to decode the bootnodes
-		bootnodes := []*peer.AddrInfo{}
+		key = libp2pKey
+	}
 
-		for _, raw := range s.config.Chain.Bootnodes {
-			node, err := StringToAddrInfo(raw)
-			if err != nil {
-				return fmt.Errorf("failed to parse bootnode %s: %w", raw, err)
-			}
+	return key, nil
+}
 
-			if node.ID == s.host.ID() {
-				s.logger.Info("Omitting bootnode with same ID as host", "id", node.ID)
+// Start starts the networking services
+func (s *Server) Start() error {
+	s.logger.Info("LibP2P server running", "addr", common.AddrInfoToString(s.AddrInfo()))
 
-				continue
-			}
+	if setupErr := s.setupIdentity(); setupErr != nil {
+		return fmt.Errorf("unable to setup identity, %w", setupErr)
+	}
 
-			bootnodes = append(bootnodes, node)
+	// Set up the peer discovery mechanism if needed
+	if !s.config.NoDiscover {
+		// Parse the bootnode data
+		if setupErr := s.setupBootnodes(); setupErr != nil {
+			return fmt.Errorf("unable to parse bootnode data, %w", setupErr)
 		}
 
-		if setupErr := s.discovery.setup(bootnodes); setupErr != nil {
+		// Setup and start the discovery service
+		if setupErr := s.setupDiscovery(); setupErr != nil {
 			return fmt.Errorf("unable to setup discovery, %w", setupErr)
 		}
 	}
 
-	go func() {
-		if err := s.runJoinWatcher(); err != nil {
-			s.logger.Error(fmt.Sprintf("Unable to start join watcher service, %v", err))
-		}
-	}()
+	go s.runDial()
+	go s.checkPeerConnections()
 
 	// watch for disconnected peers
 	s.host.Network().Notify(&network.NotifyBundle{
 		DisconnectedF: func(net network.Network, conn network.Conn) {
 			go func() {
-				s.delPeer(conn.RemotePeer())
+				s.removePeer(conn.RemotePeer())
 			}()
 		},
 	})
 
 	return nil
 }
 
+// setupDiscovery Sets up the discovery service for the node
+func (s *Server) setupDiscovery() error {
+	// Set up a fresh routing table
+	keyID := kb.ConvertPeerID(s.host.ID())
+
+	routingTable, err := kb.NewRoutingTable(
+		defaultBucketSize,
+		keyID,
+		time.Minute,
+		s.host.Peerstore(),
+		10*time.Second,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Set the PeerAdded event handler
+	routingTable.PeerAdded = func(p peer.ID) {
+		info := s.host.Peerstore().PeerInfo(p)
+		s.addToDialQueue(&info, common.PriorityRandomDial)
+	}
+
+	// Set the PeerRemoved event handler
+	routingTable.PeerRemoved = func(p peer.ID) {
+		s.dialQueue.DeleteTask(p)
+	}
+
+	// Create an instance of the discovery service
+	discoveryService := discovery.NewDiscoveryService(
+		s,
+		routingTable,
+		s.logger,
+		make(chan struct{}),
+	)
+
+	// Register a network event handler
+	if subscribeErr := s.SubscribeFn(discoveryService.HandleNetworkEvent); subscribeErr != nil {
```

Comment:
I think it's natural that Subscriber calls Subscribe method

Reply:
I completely agree, this was done in order to not have to pass jumbo data from the networking server to the discovery service when setting up the network event handler.

We will refactor this when we tackle network events 👍